mount/infoParser.go

```diff
@@ -0,0 +1,1037 @@
+		return false
+	}
+
+	infoParent := mi.GetParentMount(info)
+	if infoParent == nil {
+		return false
+	}
+
+	return info.MountPoint == infoParent.MountPoint
+}
+
+// IsBindMount verifies if the passed mountinfo entry is a bind-mount candidate.
+func (mi *mountInfoParser) IsBindMount(info *domain.MountInfo) bool {
+
+	if info == nil {
+		return false
+	}
+
+	mh := mi.service.mh
+	if mh == nil {
+		return false
+	}
+
+	devIdSlice := mi.devIdInfo[info.MajorMinorVer]
+
+	for _, elem := range devIdSlice {
+		if elem.MountID == info.MountID {
+			continue
+		}
+
+		if elem.Root == info.Root && elem.Source == info.Source {
+			return true
+		}
+	}
+
+	return false
+}
+
+// IsBindMount verifies if the passed mountinfo entry is a read-only bind-mount.
+func (mi *mountInfoParser) IsRoBindMount(info *domain.MountInfo) bool {
+
+	if info == nil {
+		return false
+	}
+
+	mh := mi.service.mh
+	if mh == nil {
+		return false
+	}
+
+	devIdSlice := mi.devIdInfo[info.MajorMinorVer]
+
+	for _, elem := range devIdSlice {
+		if elem.MountID == info.MountID {
+			continue
+		}
+
+		if elem.Root == info.Root && elem.Source == info.Source {
+			return mh.StringToFlags(elem.Options)&unix.MS_RDONLY == unix.MS_RDONLY
+		}
+	}
+
+	return false
+}
+
+func (mi *mountInfoParser) equivalentMounts(m1, m2 *domain.MountInfo) bool {
+
+	if m1 == nil && m2 == nil {
+		return true
+	} else if m1 == nil || m2 == nil {
+		return false
+	}
+
+	return m1.Root == m2.Root &&
+		m1.FsType == m2.FsType &&
+		m1.MajorMinorVer == m2.MajorMinorVer &&
+		m1.MountPoint == m2.MountPoint &&
+		m1.Source == m2.Source
+}
+
+func (mi *mountInfoParser) equivalentMountAncestors(m1, m2 *domain.MountInfo) bool {
+
+	if m1 == nil && m2 == nil {
+		return true
+	} else if m1 == nil || m2 == nil {
+		return false
+	}
+
+	p1 := m1.Mip.GetParentMount(m1)
+	if p1 != nil && m1.Mip.IsSysboxfsBaseMount(p1.MountPoint) {
+		p1 = m1.Mip.GetParentMount(p1)
+	}
+	p2 := m2.Mip.GetParentMount(m2)
+	if p2 != nil && m2.Mip.IsSysboxfsBaseMount(p2.MountPoint) {
+		p2 = m2.Mip.GetParentMount(p2)
+	}
+
+	return mi.equivalentMounts(p1, p2)
+}
+
+// IsCloneMount determines if the passed mountInfo entry is a 'clone' of an
+// entry in the sys-container's mount namespace.
+func (mi *mountInfoParser) IsCloneMount(
+	procInfo *domain.MountInfo,
+	ronlyMatch bool) bool {
+
+	mh := mi.service.mh
+	if mh == nil {
+		return false
+	}
+
+	var candidateList []*domain.MountInfo
+
+	// Extract the list of mountpoints matching the incoming process' maj-min-id.
+	devIdSlice := mi.devIdInfo[procInfo.MajorMinorVer]
+
+	for _, cntrInfo := range devIdSlice {
+		if cntrInfo.MountID == procInfo.MountID {
+			continue
+		}
+
+		// Skip readwrite candidates when operating in readonly mode.
+		if ronlyMatch && !mi.IsRoMount(cntrInfo) {
+			continue
+		}
+
+		// All candidates must meet a minimum set of criteria.
+		if cntrInfo.Root != procInfo.Root ||
+			cntrInfo.Source != procInfo.Source ||
+			mi.IsRoMount(cntrInfo) != mi.IsRoMount(procInfo) {
```

Comment:
We should check the mismatch of all flags, except the read-only flag. That's because a clone mount must match on all attributes (except read-only since we allow read-only remounting of immutable mounts).